text/0000-lwc-signals.md

```diff
@@ -0,0 +1,372 @@
+---
+title: Reactivity Model with Signals in LWC
+status: DRAFT
+created_at: 2023-11-08
+updated_at: 2023-11-08
+champion: Caridy & Dale 
+pr: https://github.com/salesforce/lwc-rfcs/pull/
+---
+
+# Reactivity Model with Signals in LWC
+
+## Summary
+
+This RFC proposes an enhancement to the reactivity model in Lightning Web Components (LWC) by introducing a Signals-based approach, inspired by similar mechanisms in libraries such as Preact. The goal is to provide a more flexible and optimized method for managing and tracking component state changes, facilitating better interaction with external data stores. The introduction of Signals aims to deprecate the current `@wire` and `@track` decorators, allowing developers to bring their own data stores and adhere to a simple protocol involving `.value`, and `.subscribe` properties. This change is designed to not only open up LWC to a broader development ecosystem but also to enable significant performance optimizations, especially when rehydrating UI components, potentially reducing the need for full component re-rendering through a more efficient subscription mechanism.
+
+## Motivation
+
+The current reactivity model in LWC relies on `@wire` and `@track` decorators, which are tightly coupled with LWC internals. This coupling leads to certain limitations:
+
+1. __Limited External Store Interaction__: LWC components cannot directly interact with external state management systems without going through LWC-specific APIs. This restricts the framework's flexibility and developer freedom to integrate with the broader JavaScript ecosystem.
+1. __Complexity and Overhead__: The use of decorators like `@wire` and `@track` introduces additional complexity and overhead in component design and state management, making it challenging for developers to reason about data flow and updates.
+1. __Performance Constraints__: The current model requires components to re-render fully to reflect state changes, which can be inefficient, especially for large and complex component trees.
+
+By adopting a Signals-based reactivity model, LWC can overcome these challenges:
+
+1. __Framework Agnosticism__: With Signals, LWC components can interact with any external data store that follows the protocol. This change would make LWC more agnostic and flexible, allowing for easier integration with various state management patterns and libraries.
+1. __Simplified State Management__: Signals provide a straightforward way to manage reactive data. By deprecating the @wire and @track decorators, we can reduce complexity, making it easier for developers to write and maintain component logic.
+1. __Performance Optimizations__: The proposed model opens up possibilities for performance optimizations by allowing selective rehydration of the UI. Components could update the DOM directly in response to state changes without a full re-render, leading to faster updates and reduced resource consumption.
```

Comment:
Nit: it's not clear to me that signals are _required_ to achieve fine-grained reactivity. E.g. the design described in https://github.com/salesforce/lwc/issues/3624 calls for adding fine-grained reactivity to template loops, without any user-facing API changes.

Also, it is entirely possible to implement signals while still re-running the entire template. It would be useful to see examples where signals unlock perf optimizations that we couldn't get otherwise.